### ISSUE TITLE
Outsource Proof Production

### ISSUE THREAD

**Issue by aszepieniec** [Neptune-Crypto/neptune-core#206]
In order to initiate a transaction, users need to produce a `SingleProof`. Presently, producing this proof is prohibitively expensive on consumer-grade hardware: it requires almost a hundred Gb of RAM and a few dozen cores.

However, the intermediate step prior to obtaining a `SingleProof` is producing a `ProofCollection`, which proves all the same things only in a more verbose way. Producing `ProofCollection`s is feasible on consumer-grade hardware.

We need to support outsourcing the last step. So if you are a lean node, you can produce the `ProofCollection` and then send it to a beefy peer. The peer produces the `SingleProof` for you.

**Basics:**
 - [x] protocol messages & handlers
 - [x] CLI argument to specify which kind of client you want to run
 - [ ] extension to peer state for lean nodes: record which peers are beefy

**DoS protections:**
 - [x] extension to peer state for beefy nodes: record time since last proof query
 - [x] (DoS protection:) limit proof query rate to *e.g.* 1 query / 10 minutes
 - [x] (DoS protection:) a peer must be connected for at least *e.g.* 10 minutes before being allowed to query a proof

Down the line we might want to consider:
 - **Pricing.** The ultimate defense against DoS attacks is to take a fee in exchange for producing a proof.
 - **Privacy.** If I were a deanonymization service, I would volunteer to produce proofs because then I get the IP and geographical location (assuming no Tor or VPN) of transaction initiators. A dandelion protocol addresses this issue.

**Comment by dan-da** [comment]
Perhaps pricing should be built in from the start.

It seems like the "beefy" proving nodes could advertise a payment address and fee amt, and then enforce a rule that they will only generate a proof if the Tx contains an output to that address with adequate fee.  The advertised fee could be 0, which effectively makes fee collection optional for the proving node operator.

**Comment by Sword-Smith** [comment]
I implemented this as a scheduled task that checks the mempool for proofs that can be upgraded. Merged and completed through 214642f4ef89723e1c0ba69feb4ff5a6d0cc6bb7, 9dd6a14e1d038691a258514820c919b7f23c122e, and  701b1306403256f970e33a2158d7f4532f68ce8d. 

It does not distinguish from where the transaction came, as this information is not currently stored in the mempool. I believe this is an OK solution for now, but we definitely want to add pricing, I would say. 

Let's add a new issue on that though.
